airflow/jobs.py

```diff
@@ -598,171 +866,475 @@ def prioritize_queued(self, session, executor, dagbag):
+            num_queued = len(task_instances)
+            self.logger.info("Figuring out tasks to run in Pool(name={pool}) "
+                             "with {open_slots} open slots and {num_queued} "
                              "task instances in queue".format(**locals()))
+
             if open_slots <= 0:
                 continue
-            tis = sorted(
-                tis, key=lambda ti: (-ti.priority_weight, ti.start_date))
-            for ti in tis:
-                if open_slots <= 0:
-                    continue
-                task = None
-                try:
-                    task = dagbag.dags[ti.dag_id].get_task(ti.task_id)
-                except:
-                    self.logger.error("Queued task {} seems gone".format(ti))
-                    session.delete(ti)
-                    session.commit()
-                    continue
 
-                if not task:
-                    continue
+            priority_sorted_task_instances = sorted(
+                task_instances, key=lambda ti: (-ti.priority_weight, ti.execution_date))
 
-                ti.task = task
+            # DAG IDs with running tasks that equal the concurrency limit of the dag
+            dag_ids_at_max_concurrency = set()
 
-                # picklin'
-                dag = dagbag.dags[ti.dag_id]
-                pickle_id = None
-                if self.do_pickle and self.executor.__class__ not in (
-                        executors.LocalExecutor,
-                        executors.SequentialExecutor):
-                    self.logger.info("Pickling DAG {}".format(dag))
-                    pickle_id = dag.pickle(session).id
+            for task_instance in priority_sorted_task_instances:
+                if open_slots <= 0:
+                    # Can't schedule any more since there are no more open slots.
+                    break
 
-                if dag.dag_id in dag_blacklist:
-                    continue
-                if dag.concurrency_reached:
-                    dag_blacklist.add(dag.dag_id)
-                    continue
-                if ti.are_dependencies_met():
-                    executor.queue_task_instance(ti, pickle_id=pickle_id)
-                    open_slots -= 1
-                else:
-                    session.delete(ti)
-                    session.commit()
+                if simple_dag_bag.is_paused(task_instance.dag_id):
+                    self.logger.info("Not executing queued {} since {} is paused"
+                                     .format(task_instance, task_instance.dag_id))
                     continue
-                ti.task = task
 
-                session.commit()
+                # Check to make sure that the task concurrency of the DAG hasn't been
+                # reached.
+                dag_id = task_instance.dag_id
+                task_concurrency_limit = simple_dag_bag.get_task_concurrency_limit(dag_id)
```

Comment:
This is new? I'm not sure if I like the approach on another concurrency limiter. We have pools, parrallelism, dag limits. What are we trying to solve and maybe we should take a step back and look at it a bire more holistically.